File: leetcode_alg/solution40.cpp
//
// Created by wangheng on 2021/6/22.
// 组合数总和Ⅱ，每个数字只能使用一次
#include "iostream"
#include "vector"
#include "vec_util.h"

using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> path;
        vector<bool> isUsed(candidates.size(), false);

        if(candidates.size()==0){
            return res;
        }
        sort(candidates.begin(), candidates.end());
        dfs(candidates, target, res, path, 0);
        return res;
    }

    void dfs(vector<int> &candidates, int target, vector<vector<int>> &res, vector<int> path, int begin){
        if(target == 0){
            res.emplace_back(path);
            return;
        }
        for(int i = begin;i<candidates.size();i++){
            if(target - candidates[i] < 0){
                break;
            }
            if(i>begin && candidates[i] == candidates[i-1]){
                continue;
            }
            path.push_back(candidates[i]);
            dfs(candidates, target - candidates[i], res, path, i+1);
            path.pop_back();
        }
    }
};

void testSolution(){
    auto s = new Solution;
    vector<int> num = {10,1,2,7,6,1,5};

    vector<vector<int>> res = s->combinationSum2(num, 8);
    for(vector<int> v:res){
        print_vec(v);
    }
}

int main(){
    testSolution();
    return 0;
}

File: golang/solution/solution40.go
package solution

import (
	"fmt"
	"sort"
)

// 组合数总和2

func combinationSum2(candidates []int, target int) [][]int {
	var (
		path   []int   = make([]int, 0)
		result [][]int = make([][]int, 0)
		dfs    func(begin int, sum int)
	)

	dfs = func(begin, sum int) {
		if sum == 0 {
			result = append(result, append([]int{}, path...))
			return
		}

		for i := begin; i < len(candidates); i++ {
			if sum-candidates[i] < 0 {
				break
			}
			if i > begin && candidates[i] == candidates[i-1] {
				continue
			}
			path = append(path, candidates[i])
			dfs(i+1, sum-candidates[i])
			path = path[:len(path)-1]
		}
	}

	sort.Ints(candidates)
	dfs(0, target)
	return result
}

func Solution40() {
	candidates := []int{10, 1, 2, 7, 6, 1, 5}
	result := combinationSum2(candidates, 8)
	fmt.Println(result)
}
